Progress Report - June 2023 (#358)
yuzu-emu/yuzu-emu.github.io: site/content/entry/yuzu-progress-report-jun-2023/index.md

+++
date = "2023-07-07T12:00:00-03:00"
title = "Progress Report June 2023"
author = "GoldenX86"
forum = 0
+++

Yuz-ers! Our devs, old and new, show no signs of slowing down. This month we have general performance and stability improvements, many graphics bug fixes, Android gets its first wave of improvements, and much more! Helmets on, let’s ride. 

<!--more--> 

Before we start, we have some important news to share before we dive into the report. 
It seems that some users have been experiencing crashes when trying to use Vulkan or change settings. 
The culprit is none other than `Overwolf`, a mod manager/overlay that comes with a faulty and outdated Vulkan layer, messing up the GPU’s Vulkan driver. 

> We strongly advise users to get rid of this piece of… software.

Author: Too politically correct. Fix it up and we'll include it.

Now that we’ve cleared that up, let’s get started.

## Illusion of Time

One of the requirements for running some commercial games in yuzu is a firmware dump, necessary for things like Mii models, or to properly render foreign scripts or button fonts.
The firmware files also include the resources needed for proper time zone support.

Thanks to the amazing work of [toastUnlimited](https://github.com/lat9nq), yuzu {{< gh-hovercard "10603" "now ships with" >}} virtually all {{< gh-hovercard "10797" "time zone data" >}} that the Switch has, without requiring a firmware dump.

In the past, yuzu would pretend that the system was in GMT, and then adjust the universal time by the local offset in the guest.
While this worked fine, it’s not very accurate to how the Switch does it.

The default “Auto” time zone setting will now send a location-accurate time zone to the guest (instead of following the old GMT rule) with the synthesised archive.
The new system also sends a correct representation of the universal time, regardless of whether there is a firmware dump or not.

> Now, while Linux had few issues with the new system, Flatpak being its usual annoying self, Windows has some special needs. There are two important things to keep in mind: 

Author: Too politically correct.

{{< single-title-imgs
    "Hey, Noah’s eyes are up here (Xenoblade Chronicles 3)"
    "./xc3bug.mp4"
    "./xc3fix.mp4"
    >}}

- Fixes the corrupted saves thumbnails in `Xenoblade Chronicles: Definitive Edition`.

{{< single-title-imgs-compare
    "That’s Dunban over there! (Xenoblade Chronicles: Definitive Edition)"
    "./picbug.png"
    "./picfix.png"
    >}}

And more.

{{< gh-hovercard "10783" "Here’s some good news for the low-RAM folks!" >}}

Since native Switch game shaders can’t be run directly on the user’s GPU, yuzu has to recompile them into an intermediate representation or IR, which is a middle format that can be optimised and then converted to a shader that users' GPU drivers can actually handle (SPIR-V/GLSL/GLASM, no, not you Metal). For performance reasons, a significant amount of memory is preallocated to store these blocks of IR, and the old implementation would use 34MB per thread, twice.

68MB per thread may not seem like a lot, right? But this allocation is made as many times as the total thread count of the CPU, SMT included. 

> Some quick maffs, and you can see how a Ryzen 3600 would gobble up 816 MB of system RAM, a Ryzen 7950X or an i9 13900 would need 2.2GB, and many phones with 8-threads SoCs would lose 544MB.

Author: Much cleaner, but too machine-like too...
Can't decide on this one.

{{< imgs
	"./af.png| Totally not the same pic"
  >}}

While digging into the AccelerateDMA code, Epicboy found an {{< gh-hovercard "10583" "incorrect check" >}} that affected Buffer < > Texture copies.
Fixing it increased performance by a whopping 1%, but as any avid gamer knows, average framerates are only half of the story.
This change also improved frametime consistency, especially in sudden spikes, which are arguably the most noticeable during gameplay.

{{< imgs
	"./frame1.png| "
  >}}

{{< imgs
	"./frame2.png| I have become frametime, the destroyer of gameplay"
  >}}

Users like to try yuzu on a wide variety of devices; one particularly interesting example is the Tegra X1 product line from NVIDIA, the same architecture that powers the actual Nintendo Switch.
Since there are drivers available for those devices, yuzu can be run on a Linux environment, but for some reason, such drivers don’t include the VK_EXT_robustness2 extension, which was mandatory up until now.

> Newcomer [mrcmunir](https://github.com/mrcmunir) decided to change this and {{< gh-hovercard "10635" "mark the extension as optional," >}} letting the most optimal ARM boards for Switch emulation be able to enjoy some good yuzu gameplay.

Reviewer: ```suggestion
Newcomer [mrcmunir](https://github.com/mrcmunir) decided to change this and {{< gh-hovercard "10635" "mark the extension as optional," >}} making the most optimal ARM boards for Switch emulation able to enjoy gameplay with yuzu.
```

> To close the graphics section, the best for last.

Reviewer: ```suggestion
To close the graphics section, we saved the best for last.
```

Thanks to the joint work of byte[] and [GPUCode](https://github.com/GPUCode), and the help from [bylaws](https://github.com/bylaws) who allowed us to use Skyline’s BCn texture decoder, yuzu can now properly advertise {{< gh-hovercard "10837" "support for Mali GPUs!" >}}

[Kept you waiting, huh.](https://www.youtube.com/watch?v=6EBuKP-uc94)

This means users of Mediatek, old Samsung Exynos, and Google Tensor CPUs [(among others)](https://en.wikipedia.org/wiki/Mali_(processor)#Implementations) can run their games on yuzu as long as they have a [G series Mali GPU](https://en.wikipedia.org/wiki/Mali_(processor)#Variants), for example the Mali-G710 on the Pixel 7 series.

{{< single-title-imgs
    "SUPER MARIO ODYSSEY and Pokémon Mystery Dungeon Rescue Team DX. Pics taken with a Pixel 7 Pro"
    "./mali1.png"
    "./mali2.png"
    >}}

This theoretically doubles the compatibility of yuzu’s Android builds, as Mali is the most common GPU family in the Android ecosystem.

{{< single-title-imgs
    "The Legend of Zelda: Skyward Sword HD, Advance Wars 1+2: Re-Boot Camp, and OCTOPATH TRAVELER. Pics taken with a Pixel 7 Pro"
    "./mali3.png"
    "./mali4.png"
    "./mali5.png"
    >}}

> Further testing with Mali revealed that while the hardware briefly had support for the VK_EXT_extended_dynamic_state2 extension in older Mali drivers, the implementation is actually completely broken.

Reviewer: ```suggestion
Further testing with Mali revealed that while the driver implemented support for the VK_EXT_extended_dynamic_state2 extension in older Mali drivers, the implementation for VK_EXT_extended_dynamic_state was not fully functional, and yuzu was not behaving correctly in this case.
```

> This is most likely the reason the extension was removed from current drivers. 

Reviewer: ```suggestion
This is most likely the reason both extensions were removed by Google from Pixel drivers. 
```

> To avoid users of affected drivers from suffering random crashes, byte[] {{< gh-hovercard "10790" "disabled the use of the extension" >}} on Mali hardware.

Reviewer: ```suggestion
To avoid users of affected drivers from getting no rendering output at all, byte[] {{< gh-hovercard "10790" "disabled the use of the extension" >}} on Mali hardware.
```

It should be noted that while the Mali drivers are much better at following the Vulkan specification than others, the physical hardware specifications are lacking.
Only high end devices will reach playable framerates at the moment. 
Lower end Mali products will be too weak for Switch emulation, even if they meet the driver requirements to run the emulator.
But hey, you’re welcome to try it out and see how it goes!

> There’s also a LOT of work put into the Android user interface. 

Reviewer: ```suggestion
There’s also been a LOT of work put into the Android user interface. 
```

What can we say–we’re new at this.

> To begin with, [PabloG02](https://github.com/PabloG02) is giving us a hand again, this time by {{< gh-hovercard "10551" "storing the position of overlay controls" >}} as a percentage, so if there is a change in the dimensions of the window, for example by using a foldable device, or rotating the screen on a tablet, the overlay controls stay in approximately the same location.

Reviewer: ```suggestion
To begin with, [PabloG02](https://github.com/PabloG02) has given us a hand again, this time by {{< gh-hovercard "10551" "storing the position of overlay controls" >}} as a percentage, so if there is a change in the dimensions of the window, for example by using a foldable device, or rotating the screen on a tablet, the overlay controls stay in approximately the same location.
```

We need to do this to ensure users don’t lose their progress, but it can also be used to include some cool features.

One possibility is support for Picture-in-Picture when minimising the app.
t895 did the {{< gh-hovercard "10633" "preliminary work" >}} to allow changing the aspect ratio, and newcomer [AbandonedCart](https://github.com/AbandonedCart) {{< gh-hovercard "10639" "did the actual magic." >}}

{{< single-title-imgs
    "That’s a spicy music player (Super Smash Bros. Ultimate)"
    "./pip1.png"
    "./pip2.png"
    >}}

Later, a  {{< gh-hovercard "10811" "button to mute/unmute" >}} was added to the Picture-in-Picture overlay.

Android doesn’t only run on phones and tablets. You can find people crazy enough to play yuzu on TVs too!
For those madlads, newcomer [qurious-pixel](https://github.com/qurious-pixel) added support for an {{< gh-hovercard "10650" "Android TV" >}} banner.
Thank you!

{{< imgs
	"./tv.png| NVIDIA SHIELD anyone?"
  >}}

Since release, users have reported games looking strangely blurry, and issues with the aspect ratio of the rendering window and the orientation of the overlay controls when flipping the device.

> This is a bit of an embarrassing one: all those issues were caused by decoupling the rotation of the surface with the frontend layer.

Reviewer: ```suggestion
This is a bit of an embarrassing one: all those issues were caused by decoupling the rotation of the Vulkan surface with the frontend layer.
```

  >}}

An often forgotten feature of the Nintendo Switch is touch support. 
While in handheld mode, some games allow you to play without controls, just with the screen. 
Support was present since release for the Android build, but it wasn’t accurate–the touch event would happen offset to the side of where the user would touch their finger. german77 fixed this {{< gh-hovercard "10751" "touchy behaviour," >}} making playing visual novels with no overlay buttons a joy.

One of the nice additions of Android 13 is the option to set {{< gh-hovercard "10760" "per-app languages," >}} so properly declaring the available language translations allows users to take advantage of this feature. 
Thanks [flTobi](https://github.com/FearlessTobi)!

{{< imgs
	"./lang.png| Your phone knows more languages than you"
  >}}

{{< gh-hovercard "10808" "Expanding the Debug section" >}} a bit, t895 added the option to disable `CPU Fastmem`.
This will reduce performance, but is useful for debugging purposes.

{{< imgs
	"./debug.png| Extremely recommended to leave at default values"
  >}}

Additionally, an option to select the audio backend was added too.

{{< imgs
	"./audio.png| Volume controls"
  >}}

> Because the option is only supported on Qualcomm devices, t895 {{< gh-hovercard "10864" "now hides" >}} the `Install GPU driver` option if the user is running a Mali or Intel/AMD x86 device.

Reviewer: ```suggestion
Because it is only supported on Qualcomm devices, t895 {{< gh-hovercard "10864" "now hides" >}} the `Install GPU driver` option if the user is running a non-Qualcomm device.
```

While simulating a power limited scenario (capping the PPT to 40W), performance improved up to 60%, but on average a 20% gain is more often expected.

Now we can guess what some of you may be thinking, “how much does this help the Steam Deck? That’s a power-limited system with only four cores!” 
It makes no difference on the Deck. Let us explain.

The Steam Deck by default runs a Linux distribution. 
Linux doesn’t benefit from these changes as its kernel, unlike Windows’, is actually decent and can run high-precision timer events without any problem or undocumented SDK usage. If there’s no need to spin the CPU at all, it gets to idle properly. 
So unless you installed Windows onto your Steam Deck, you already had the best performance it can offer.

Long gameplay sessions can sometimes feel like a gamble regarding stability. 
That’s a very common problem for emulators due to how they have to handle memory in unique ways to match two completely different systems. 
This was especially true for Linux users, who had to rely on increasing the size of the `vm.max_map_count` kernel parameter for certain games to avoid out-of-memory crashes. 

> Since the emulator needs to create placeholder memory mappings to keep up with the virtual memory requirements of the game, it wasn’t hard to saturate the default value, resulting in a crash once the placeholder mappings outnumber the max mapping count.

Reviewer: ```suggestion
Since the emulator needs to create placeholder memory mappings to keep up with the virtual memory requirements of the game, it wasn’t hard to saturate the default value, resulting in a crash once the placeholder mappings outnumbered the max mapping count.
```

This will work for most 64-bit games, but 32-bit Switch games have some special requirements that complicate this.

NCE is a project that will take some time, but for now, byte[] already implemented a way to {{< gh-hovercard "10747" "decouple Dynarmic" >}} from the ARM interface, allowing for a separate CPU backend in the future.
Let’s see what the future brings.

We’re still calling it Project Nice, right? Right.

Regarding 32-bit games, it’s worth mentioning that behind the scenes, [Merry](https://github.com/merryhime) constantly works on optimising and fixing bugs in Dynarmic.
Recently, Merry was able to {{< gh-hovercard "10933" "enable an optimization" >}} for ARM64 host hardware. 
Block linking allows blocks of guest code that directly jump to each other to directly jump to each other when recompiled as well.
After enabling it, this resulted in a 60-70% performance boost for 32-bit games like `Mario Kart 8 Deluxe` on Android SoCs.
This serves as yet another reminder that emulation is very CPU-focused.

## Input improvements

The work on making those fancy plastics with Near-Field Communication (NFC) behave like on the Switch continues.

> With the proper implementation finished [last month](https://yuzu-emu.org/entry/yuzu-progress-report-may-2023/#input-and-amiibo-improvements), german77 can now focus on the last items on the checklist.

Reviewer: ```suggestion
With the proper implementation finished [last month](https://yuzu-emu.org/entry/yuzu-progress-report-may-2023/#input-and-amiibo-improvements), german77 has focused on the last items on the checklist.
```

On the Switch, Amiibo data is stored in the console every time data is loaded or saved. 
This is intended to help games restore corrupted Amiibo tags in case of errors.
{{< gh-hovercard "10623" "With this implementation," >}} yuzu now can do the same with Joy-Cons and Pro Controllers.
The backups will be stored in `%appdata%/amiibo/backup`.

Our input dev added support for Foomiibos, blank but configurable Amiibo dumps with a signature included at the end, by {{< gh-hovercard "10795" "adding their size as a valid input." >}}

Activision releases NFC equipped toys (aka not-Amiibos) for the `Skylanders` games saga, and they are compatible with the Nintendo Switch, so of course, they must work on yuzu too! 
{{< gh-hovercard "10842" "The implementation" >}} should be transparent; just connect and map a controller, and it’s ready to scan the toys.

And to finish up the NFC section, {{< gh-hovercard "10903" "bugfixes and support for third-party controllers" >}} in a single pull request to rule them all.

Controllers sometimes fail to properly restore to a previous state when pairing. 
This could make yuzu lose input when the controller fails to use the NFC sensor.

> So, if the game stops polling for Amiibos while there’s one in range, the Amiibo would be lost but not signaled.

Reviewer: ```suggestion
If the game stopped polling for Amiibos while one was in range, the Amiibo would be lost but not signaled.
```

This pull request addressed this particular scenario.

Many third-party controllers don’t accept all commands. 
Reducing the number of attempts to obtain valid data and disabling most commands if errors occur will minimise the stuttering caused by the configuration process when setting up the NFC interface.

Lastly, if the controller properly reports support for NFC, but the user loads an amiibo from a binary dump, we would mistakenly show an error. 
This is now fixed.

Now, what about regular input changes? Well, we have a bit to talk about there too.

SDL, the external library used here to handle input (it can do much more than that), does get updates, and it doesn’t hurt to keep up with those from time to time, after validating they are stable, of course.

> {{< gh-hovercard "10873" "Updating to version 2.28.0" >}}improved the calibration profile used on official Switch controllers, added support for the new Dual Sense Edge controller from Sony, {{< gh-hovercard "10891" "fixed some issues with Pro Controllers," >}} and added other minor changes and fixes that benefit the emulator.

Reviewer: ```suggestion
{{< gh-hovercard "10873" "Updating to version 2.28.0" >}}improved the calibration profile used on official Switch controllers, added support for the new DualSense Edge controller from Sony, {{< gh-hovercard "10891" "fixed some issues with Pro Controllers," >}} and added other minor changes and fixes that benefit the emulator.
```

A {{< gh-hovercard "10588" "memory cache" >}} was added to fully realise the load time benefits on Microsoft’s OS too.

In another single-line-of-code revelation, byte[] discovered that {{< gh-hovercard "10718" "increasing the size of the buffer" >}} when copying files can more than triple the installation speed of updates and DLCs. 
Hurray for simple fixes like this!

A common problem modders and users had to deal with is a hard limit on the number of files that can be loaded per session.
Games like `Fire Emblem Engage` have so many files that it previously was not even possible to replace all of them with a mod.
Of course, byte[] wasn’t happy with this, so by adding a file [LRU](https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_recently_used_(LRU)) cache to control the number of open files, he {{< gh-hovercard "10729" "removed the file limit." >}}
Users can now run as many mods as they want on their games, and they can have as many files as they need too!

But the fun doesn’t stop there for byte[]. Yet again, Windows demands special attention…
{{< gh-hovercard "10806" "A couple of optimizations," >}} and Windows users get to properly benefit too.

> An oversight in the code responsible for caching the installed mods disabled the entire implementation, making yuzu take considerable time when booting a game with mods.

Reviewer: ```suggestion
An oversight in the code responsible for caching RomFS mods prevented any caching from taking effect at all, making yuzu take typically twice as long as it needed to when loading a game with them enabled.
```

By {{< gh-hovercard "10594" "fixing" >}} this oversight, byte[] further reduced game load times with mods.

## Audio fixes

Linux users running the SDL audio backend reported having no audio out when using the JACK output.

> The problem, as newcomer [zeltermann](https://github.com/zeltermann) found out after a good investigation, is in how SDL identifies the CPU characteristics of the user’s hardware.

Reviewer: ```suggestion
The problem, as newcomer [zeltermann](https://github.com/zeltermann) found out after some investigation, is in how SDL identifies the CPU characteristics of the user’s hardware.
```

> SDL’s `CPUInfo` subsystem is called to report the specifications of the user’s CPU, SDL then uses that information to properly configure itself for the environment it will run on, setting optimizations and deciding which internal code paths to use.

Reviewer: ```suggestion
SDL’s `CPUInfo` subsystem is called to report the specifications of the user’s CPU. SDL then uses that information to properly configure itself for the environment it will run on, setting optimizations and deciding which internal code paths to use.
```

[SIMD](https://en.wikipedia.org/wiki/Single_instruction,_multiple_data) support is reported as disabled, but [SSE2](https://en.wikipedia.org/wiki/SSE2) is assumed as enabled.
This contradiction leads to the fallbacks intended to run in case SIMD support was disabled to fail, making some functions return null, completely destroying audio output.

This chain reaction of events is solved by giving SDL what it wants: a proper {{< gh-hovercard "10739" "report by CPUInfo." >}}
Now Linux users can enjoy their games with audio when using the SDL output.
Thank you!

Out-of-bounds, the silent killer of so many programs… yuzu is of course not immune to them, as Morph found out.
An {{< gh-hovercard "10966" "out-of-bounds write" >}} leads to corruption in an audio heap buffer, which leads to beeping noises, speaker rattling, and crashes in multiple games, including but not limited to `Xenoblade Chronicles: Definitive Edition`, `Xenoblade Chronicles 2`, `Super Smash Bros. Ultimate`, and more.

A few tweaks here and there in the audio core code, and the issues are gone.

## UI changes

Switch games provide pretty game icons for the launcher. 

> Needless to say, yuzu couldn’t let the opportunity to make good use of them on the game list go. 

Reviewer: ```suggestion
Needless to say, yuzu makes good use of them while rendering the game list. 
```

What wasn’t considered when using these icons for the user interface is that some games offer multiple alternatives for different languages. 
yuzu would just pick the first one available.

Enter newcomer [keve1227](https://github.com/keve1227) who coded in the {{< gh-hovercard "10591" "logic required" >}} to properly pick the right image depending on the console language the user picks in `Emulation > Configure… > System > Language`.
Thanks!

{{< single-title-imgs
    "Gotta find’em all"
    "./langbug.png"
    "./langfix.png"
    >}}

The OpenGL API has three available shader backends in yuzu, and since the only way to know which one is selected is from checking the option in `Emulation > Configure… > Graphics`, newcomer [xcfrg](https://github.com/xcfrg) decided it was due time to {{< gh-hovercard "10614" "add this info" >}} to the status bar.
Thanks!

{{< imgs
	"./oglui.png| You won’t forget you left GLASM enabled any more!"
  >}}

## Hardware section

> Before starting with the driver rants, we want to share a possible way to save up to 2 GB of VRAM by taking advantage of Windows 11’s per-app graphics selector. 

Reviewer: ```suggestion
Before starting with the driver rants, we want to share a possible way to save up to 2 GB of VRAM by taking advantage of Windows 11’s per-app graphics selector, assuming you have multiple GPUs available in your system. 
```

To do this, we will take advantage of some relatively recent additions to Windows 11.

First step after setting up the GPUs (desktop PCs may need to enable integrated GPUs on the UEFI configuration/BIOS) is to select the high performance card on display settings:

{{< single-title-imgs
    "On desktop PCs, Windows 10 fails here by not letting you select which card is which"
    "./gpu1.png"
    "./gpu2.png"
    "./gpu3.png"
    >}}

While this is enough for yuzu after a reboot, you will also need to manually set up each game or program that requires using the high performance GPU:

{{< single-title-imgs
    "Tech tip: Forza is faster if you run it with your main GPU"
    "./gpu4.png"
    "./gpu5.png"
    "./gpu6.png"
    >}}

Like this, you get the benefits of more free VRAM on the high performance GPU, and if the secondary GPU is more efficient, a power consumption reduction overall when using the PC. More available VRAM results in improved minimum framerates (the 1% low for example).

The cons for using this method is that frame data is transferred over the PCIe interface, which can be saturated. 

> This would result in lower average framerates and more input lag.

Reviewer: ```suggestion
This can result in slightly lower average framerates and more input lag.
```

### AMD

Since driver release 23.7.1, a mysterious older “23.10.01.41 for Additional Vulkan Extensions” version, and the beta driver only available to Windows Insider testers, AMD official Vulkan drivers introduced a regression in how the VK_EXT_extended_dynamic_state3 extension handles colour blending.
[Your writer](https://github.com/goldenx86) {{< gh-hovercard "10946" "blocked the affected bit" >}} while we monitor if future AMD driver releases manage to solve this problem.

{{< single-title-imgs-compare
	"Who gave Link the Batman’s Detective Mode? (The Legend of Zelda: Tears of the Kingdom)"
	"./amdbug.png"
	"./amdfix.png"
>}}

The performance loss from this change is very small, so feel free to update to the latest drivers.

[GPUCode](https://github.com/GPUCode) has found a way to emulate D24 support on AMD GPUs by using the `VK_EXT_depth_bias_control` extension, which is currently only available on the RADV Mesa driver on Linux.
In the near future, we’ll check if this method solves the remaining graphical issues specific to AMD hardware. 
If it works, it would also solve the issue on the AMD official drivers, once/if the extension is supported there too.

> We know it’s not fun having to switch to OpenGL to be able to play some games, we’re ready to combat even hardware limitations to defeat this foe.

Reviewer: ```suggestion
We know it’s not fun having to switch to OpenGL to be able to play some games, so we’re gearing up to combat even hardware limitations to defeat this foe.
```

### Intel

> Intel got its fair share of fixes this month too, first of all, as promised last month, byte[] added a translation for FP64 shaders, to turn them into FP32 ones.

Reviewer: ```suggestion
Intel got its fair share of fixes this month too.
First of all, as promised last month, byte[] added a translation for FP64 use in shaders, to turn uses into FP32.
```

This was required because Generation 12 Intel graphics (UHD 700/Iris Xe/Arc A series) removed support on a hardware level for double precision operations.

> With this {{< gh-hovercard "10693" "shader translation" >}} in place, games will no longer suddenly crash when trying to run very high precision shaders, for example during the opening cutscene of `Tears of the Kingdom` when starting a new game.

Reviewer: ```suggestion
With this {{< gh-hovercard "10693" "shader translation" >}} in place, games will no longer suddenly crash when trying to use these shaders, for example during the opening cutscene of `Tears of the Kingdom` when starting a new game.
```

> The next issue is an interesting mix of unique behaviour by the Nintendo Switch, yuzu’s code at fault, and another case of lack of hardware support. 

Reviewer: ```suggestion
The next issue is an interesting mix of unique behaviour by the Switch's own graphics drivers, yuzu’s code at fault, and another case of lack of hardware support. 
```

> Intel defined the use of barriers as specified by SPIR-V, but the Switch’s Tegra X1 (and interestingly, any other GPU vendor besides Intel, including Android ones) prefers to follow the PTX ISA documentation for barriers, which defines what to do if no checks for existing threads are done.

Reviewer: ```suggestion
Intel's GPUs allow the use of control barriers as specified by SPIR-V, which requires all threads in a dispatch to execute the same barriers at the same times, and don't allow any threads to exit early if another barrier will occur.
But all other GPU vendors besides Intel, including mobile GPUs, do allow threads to exit early, and will free up barriers on the remaining threads as needed.
This situation caused Intel GPUs to hit device losses, as guest compute shaders had barriers that ended up hitting this exact type of control flow.
```

> By {{< gh-hovercard "10699" "removing barriers" >}} in the conditional control flow, Intel GPUs are free from their device loss crashes.

Reviewer: ```suggestion
By {{< gh-hovercard "10699" "removing barriers" >}} after conditional control flow, Intel GPUs are free from their device loss crashes.
```

And lastly, one loose end remained, our toggle to disable the compute pipeline on Intel GPUs.

> Since this option was a stopgap solution until a driver update was released, it has become irrelevant now that Intel took their long time to release a proper fix, so toastUnlimited made the option only show up in yuzu’s settings if an old driver known to be affected is still being run by the user.

Reviewer: ```suggestion
This option was a stopgap solution until a driver update was released. But since Intel has finally published a driver update fixing the issues, it has now become irrelevant. toastUnlimited therefore modified the option to only show up in yuzu’s settings if an old driver known to be affected is still in use.
```

[Remember to update your drivers!](https://www.intel.com/content/www/us/en/download/729157/intel-arc-iris-xe-graphics-beta-windows.html)

With all driver and yuzu stability-related issues fixed, we can finally recommend Intel hardware again, especially current-generation integrated GPUs like the Iris Xe. 
While the lower end 80EU-equipped Iris Xe can bottleneck performance at docked 1X mode in the most intensive games, it makes up for that by having zero texture related stuttering thanks to being the last breed of Intel GPUs with native ASTC texture support. 
We don’t know if future iGPUs will keep supporting ASTC, but it’s a reality that Arc dedicated GPUs don’t offer it.

Still, we can finally focus on working on the few remaining quirks for the Generation 12 architecture.
Right now, the Iris Xe can easily beat a Geforce MX450 in ASTC intensive games like `The Legend of Zelda: Tears of the Kingdom`, using less RAM and having no texture stuttering. 
Here, the E-cores are a blessing for shader building.

> Hopefully future Intel products offer better FP16, higher performance, longer driver life support, don’t base their pricing on software gimmicks like their competition, and add a much-needed separate compute queue.

Reviewer: ```suggestion
Hopefully future Intel products offer better FP16, higher performance, longer driver life support, don’t base their pricing on software gimmicks like their competition, and add a much-needed dedicated compute queue.
```

We can’t work on fixing regressions on a driver that is a work in progress; the only thing to do is to wait for it to mature.

#### Mali, good drivers, slow hardware, weird decisions

Mali is the complete opposite of Adreno. While there is no current Mesa support for it, we found their proprietary Vulkan driver to be in great shape once its initial quirks were addressed.
The same can’t be said about their OpenGL ES driver as our friends at Citra will gladly tell you, but thankfully that’s not an issue here.
The main problem is that the most common hardware the driver gets to run on is pretty subpar, at least for Switch emulation.
Clearly, high-end devices won’t have much issue in the GPU front as they are not far behind Adreno in hardware specs, but the rest of the product stack will suffer from low framerates and GPU bottlenecks.

Unfortunately, due to an unstable kernel API, and the lack of a maintained Mesa driver for these devices, there isn't currently any practical way to replace the running driver on a Mali device with a better one.
If a device ships with an outdated driver too old for yuzu and the vendor doesn’t update it, chances are nothing can be done about it.

> That’s planned obsolescence at its finest, or sheer incompetence. We’ll let you be the judge of that.

Reviewer: ```suggestion
That’s either planned obsolescence at its finest, or sheer incompetence. We’ll let you be the judge of that.
```

The results should be better with a proper mobile platform, but alas, we can only show what we tested.

Power consumption got a 10-16% reduction, which translates with this particular CPU model to around a 10W reduction.
While simulating a power limited scenario (capping the PPT to 40W), performance improved up to 60%, but on average a 20% gain is more often expected.

Now we can guess what some of you may be thinking, “how much does this help the Steam Deck? That’s a power-limited system with only four cores!” 
It makes no difference on the Deck. Let us explain.

The Steam Deck by default runs a Linux distribution. 
Linux doesn’t benefit from these changes as its kernel, unlike Windows’, is actually decent and can run high-precision timer events without any problem or undocumented SDK usage. If there’s no need to spin the CPU at all, it gets to idle properly. 
So unless you installed Windows onto your Steam Deck, you already had the best performance it can offer.

> Long gameplay sessions can sometimes feel like a gamble regarding stability. 

Reviewer: random linux section stuck in the middle of an unrelated header

Author: Making a kernel section for just this would be a bit sad innit

Reviewer: All I'm saying is, it definitely doesn't belong here. Maybe a new section is or isn't warranted, but this feels extremely out of place

Author: Let me see what I can cook

In order to avoid crashes in Unreal Engine 4 games like this, the LLE slow code path will be used instead.

{{< imgs
	"./pm4.png| More proof that games don’t need high requirements to be fun (Pikmin 4 Demo)"
  >}}

This pull request introduced a regression that affected rendering in `Metroid Prime Remastered`.
Thankfully, [gidoly](https://github.com/gidoly) quickly solved it by {{< gh-hovercard "11012" "merging both approaches," >}} old and new.

{{< single-title-imgs-compare
	"Chozo Suite error 404: Texture not found (Metroid Prime Remastered)"
	"./mprbug.png"
	"./mprfix.png"
>}}

While working on these crashes, byte[] changed the behaviour of the Vulkan memory manager to prefer (instead of require) {{< gh-hovercard "10994" "using device local memory" >}} (VRAM) for image memory, which ends up allowing up to 50% of shared memory (system RAM) to be used by the GPU.
This simple trick makes most Unreal Engine 4 games stable, but it is very likely it has the cost of making VRAM intensive games stutter more often once VRAM gets close to full, for example, when running `The Legend of Zelda: Tears of the Kingdom` on a GPU with 4GB of VRAM or less.

> It’s a high price to pay considering Tears of the Kingdom is by far the most played game on yuzu, but we think the huge number of games that benefit from stable gameplay outweighs the latest Zelda stuttering a bit more on low end hardware.

Reviewer 2: ```suggestion
It’s a high price to pay considering `Tears of the Kingdom` is by far the most played game on yuzu, but we think the huge number of games that benefit from stable gameplay outweighs the latest Zelda stuttering a bit more on low-end hardware.
```

By {{< gh-hovercard "10699" "removing barriers" >}} after conditional control flow, Intel GPUs are free from their device loss crashes.

And lastly, one loose end remained, our toggle to disable the compute pipeline on Intel GPUs.
This option was a stopgap solution until a driver update was released. But since Intel has finally published a driver update fixing the issues, it has now become irrelevant. toastUnlimited therefore modified the option to only show up in yuzu’s settings if an old driver known to be affected is still in use.
The disable compute pipeline toggle is now a {{< gh-hovercard "10835" "legendary drop option" >}} only available for the most unlucky Generation 12 users, as the option will also remain hidden for unaffected Generation 11 and older devices.
[Remember to update your drivers!](https://www.intel.com/content/www/us/en/download/729157/intel-arc-iris-xe-graphics-beta-windows.html)

With all driver and yuzu stability-related issues fixed, we can finally recommend Intel hardware again, especially current-generation integrated GPUs like the Iris Xe. 

> While the lower end 80EU-equipped Iris Xe can bottleneck performance at docked 1X mode in the most intensive games, it makes up for that by having zero texture related stuttering thanks to being the last breed of Intel GPUs with native ASTC texture support. 

Reviewer 2: ```suggestion
While the lower-end 80EU-equipped Iris Xe can bottleneck performance at docked 1X mode in the most intensive games, it makes up for that by having zero texture related stuttering thanks to being the last breed of Intel GPUs with native ASTC texture support. 
```

+++
date = "2023-07-07T12:00:00-03:00"
title = "Progress Report June 2023"
author = "GoldenX86"
forum = 0
+++

Yuz-ers! Our devs, old and new, show no signs of slowing down. This month we have general performance and stability improvements, many graphics bug fixes, Android gets its first wave of improvements, and much more! Helmets on, let’s ride. 

<!--more--> 

Before we start, we have some important news to share before we dive into the report. 
It seems that some users have been experiencing crashes when trying to use Vulkan or change settings. 
The culprit is none other than `Overwolf`, a mod manager/overlay that comes with a faulty and outdated Vulkan layer, messing up the GPU’s Vulkan driver. 
We strongly advise users to get rid of this piece of… software.

Now that we’ve cleared that up, let’s get started.

> ## Illusion of Time

Author: Illusion of Time is a game.

{{< imgs
	"./af.png| Anisotropic Filtering at 16x may affect performance on integrated GPUs, something to keep in mind"
  >}}

In another batch of old issues fixed by Blinkhawk, he managed to pinpoint the cause for Unreal Engine 4 games having broken textures and crashing when running with Vulkan, a problem that mostly affected NVIDIA cards.

Unreal Engine 4 is a big fan of [sparse textures](https://docs.unity3d.com/Manual/SparseTextures.html) to the point of really pushing the limits of the texture cache code.
A bug affecting how NVN virtual images were being remapped after sparse memory gets converted to regular mapped memory caused games like `SHIN MEGAMI TENSEI V`, `Bravely Default`, the `Pikmin 4 Demo`, and many others to display corrupted textures at random.

{{< single-title-imgs-compare
    "Optical camouflage gone wrong, clearly Protag-kun here is not as good as major Motoko (SHIN MEGAMI TENSEI V)"
    "./yvcbug1.png"
    "./yfcfix1.png"
    >}}

{{< single-title-imgs-compare
    "It's not a phase, mom! (BRAVELY DEFAULT II)"
    "./yfcbug1.png"
    "./yfcfix2.png"

>     >}}

Author: Will cut them down. The site's autoresize is a pain to deal with.

Not satisfied with that, byte[] solved the problem by {{< gh-hovercard "10670" "using a higher colour bit depth." >}}
Sometimes, solutions don’t need to be complicated.

{{< single-title-imgs-compare
	"For those old enough to remember what it felt getting a GPU that could do 24-bit colour (SUPER MARIO ODYSSEY)"
	"./fxaabug.png"
	"./fxaafix.png"
>}}

Many Switch games support dynamic framerates, or get mods that support dynamic framerates, meaning unlocking the framerate (by default by pressing Ctrl + U) is a great way to get a much better and smoother experience while emulating those games.
By the way, here’s a [list](https://github.com/Law022/Dynamic-Frame-Rate/wiki/Game--List) of such games for those curious.

One unwanted side effect with this is that video playback won’t be able to keep up with the increased framerate, often desynchronizing audio while still having to wait for the video to end.
To mitigate this, byte[] {{< gh-hovercard "10666" "added a toggle" >}} in `Emulation > Configure… > Graphics > Advanced` called “Sync to framerate of video playback” that, when enabled, will dynamically re-enable the framerate limit while a video cutscene is playing.

{{< imgs

> 	"./af.png| Totally not the same pic"

Author: Images comments don't support formatting.

While the lower end 80EU-equipped Iris Xe can bottleneck performance at docked 1X mode in the most intensive games, it makes up for that by having zero texture related stuttering thanks to being the last breed of Intel GPUs with native ASTC texture support. 
We don’t know if future iGPUs will keep supporting ASTC, but it’s a reality that Arc dedicated GPUs don’t offer it.

Still, we can finally focus on working on the few remaining quirks for the Generation 12 architecture.
Right now, the Iris Xe can easily beat a Geforce MX450 in ASTC intensive games like `The Legend of Zelda: Tears of the Kingdom`, using less RAM and having no texture stuttering. 
Here, the E-cores are a blessing for shader building.

Hopefully future Intel products offer better FP16, higher performance, longer driver life support, don’t base their pricing on software gimmicks like their competition, and add a much-needed dedicated compute queue.

### Android

With the release of the Android builds, the whole team is learning about the platform’s ecosystem, its features and limitations, quirks and advantages.
There are fun times ahead for sure.

> This means we also got to learn about the depressing state of its GPU drivers…

Reviewer 2: ```suggestion
This means we also learned about the depressing state of its GPU drivers…
```

#### Adreno, or just waiting for Mesa

Adreno hardware has great specifications, but their proprietary official drivers are incredibly bad!
We’re starting to work on improving rendering on the official Adreno drivers used by default on Qualcomm equipped devices, but there’s only so much we can do.

> The real solution comes from, you guessed it, the team that manages to make good drivers with no access to confidential information, Mesa.

Reviewer 2: ```suggestion
The real solution comes from, you guessed it, the team that manages to make good drivers with no access to confidential information: Mesa.
```

One has to wonder why Android vendors don’t just use Mesa to begin with. Nothing deemed “confidential” or “trade-secret” here is more important than providing good software support! 

While the Adreno 600 series is well-supported and has great rendering with Mesa Turnip drivers, the 700 series is too new to enjoy the same benefits.
There have been a few packaged releases by [K11MCH1](https://github.com/K11MCH1) using the currently-in-development a700 branch of Mesa Turnip, but it’s exactly what it says on the tin: currently in development.
We can’t work on fixing regressions on a driver that is a work in progress; the only thing to do is to wait for it to mature.

#### Mali, good drivers, slow hardware, weird decisions

Mali is the complete opposite of Adreno. While there is no current Mesa support for it, we found their proprietary Vulkan driver to be in great shape once its initial quirks were addressed.
The same can’t be said about their OpenGL ES driver as our friends at Citra will gladly tell you, but thankfully that’s not an issue here.
The main problem is that the most common hardware the driver gets to run on is pretty subpar, at least for Switch emulation.

> Clearly, high-end devices won’t have much issue in the GPU front as they are not far behind Adreno in hardware specs, but the rest of the product stack will suffer from low framerates and GPU bottlenecks.

Reviewer 2: ```suggestion
Clearly, high-end devices won’t have much issue on the GPU front as they are not far behind Adreno in hardware specs, but the rest of the product stack will suffer from low framerates and GPU bottlenecks.
```
